src/plugins/chatview/utils.js

```diff
@@ -71,3 +73,120 @@ export function resetElementHeight (ev) {
         ev.target.style = '';
     }
 }
+
+
+/**
+ * Handle XEP-0147 "query actions" invoked via xmpp: URIs.
+ * Supports message sending, roster management, and future actions.
+ *
+ * Example URIs:
+ *   xmpp:user@example.com?action=message&body=Hello
+ *   xmpp:user@example.com?action=add-roster&name=John&group=Friends
+ */
+export async function routeToQueryAction(event) {
+    const { u } = _converse.env;
+
+    try {
+        const uri = extractXMPPURI(event);
+        if (!uri) return;
+
+        const { jid, queryParams } = parseXMPPURI(uri);
+        if (!u.isValidJID(jid)) {
+            return log.warn(`routeToQueryAction: Invalid JID: "${jid}"`);
+        }
+
+        const action = queryParams.get('action');
+        if (!action) {
+            log.debug(`routeToQueryAction: No action specified, opening chat for "${jid}"`);
+            return api.chats.open(jid);
+        }
+
+        switch (action) {
+            case 'message':
+                await handleMessageAction(jid, queryParams);
+                break;
+
+            case 'add-roster':
+                await handleRosterAction(jid, queryParams);
+                break;
+
+            default:
+                log.warn(`routeToQueryAction: Unsupported XEP-0147 action: "${action}"`);
+                await api.chats.open(jid);
+        }
+    } catch (error) {
+        log.error('Failed to process XMPP query action:', error);
+    }
+}
+
+/**
+ * Extracts and decodes the xmpp: URI from the window location or hash.
+ */
+function extractXMPPURI(event) {
+    let uri = null;
+
+    // Case 1: protocol handler (?uri=...)
+    const searchParams = new URLSearchParams(window.location.search);
```

Comment:
In this repo, we use `snake_case` to name variables (unless the variable refers to a function) and `camelCase` for functions.

I know this is not how most other JavaScript projects name their variables, but this is an old convention dating back to 2013, when I was influenced by the naming conventions in Python.

To avoid confusion and inconsistency, please stick to the convention of using `snake_case` for variables.